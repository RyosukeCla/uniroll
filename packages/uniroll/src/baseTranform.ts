import { transform as transformBabel, TransformOptions } from "@babel/core";

export function createTransformerx(options: TransformOptions = {}) {
  return async (code: string, id: string) => {
    const ret = transformBabel(code, {
      filename: id,
      ...options,
    });
    return ret && (ret.code as string);
  };
}
