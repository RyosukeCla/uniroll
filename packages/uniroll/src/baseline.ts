import { Options } from "..";
import { rollup } from "rollup";
import path from "path";
import { memfsPlugin } from "rollup-plugin-memfs";

export async function baseline(options: Options & { fs: any }) {
  const fs = options.fs;

  const input = options.useInMemory
    ? path.join("/", options.input)
    : path.join(options.cwd, options.input);

  const bundle = await rollup({
    input,
    plugins: [...(options.rollupPlugins || []), memfsPlugin(fs)],
  });
  return bundle;
}
