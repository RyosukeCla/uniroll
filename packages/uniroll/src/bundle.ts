import { virtualFs } from "rollup-plugin-virtual-fs";
import { httpResolve } from "rollup-plugin-http-resolve";
import replace from "@rollup/plugin-replace";
import workerPlugin from "./plugins/worker-plugin";
import {
  createImportMapsFallback,
  defaultCache,
  defaultCompilerOptions,
  defaultDefine,
  defaultImportMaps,
  transform,
} from "./shared";
import { rollup } from "rollup";
import json from "@rollup/plugin-json";
import { CompileOptions } from "./types";

export function bundle({
  input,
  files,
  cache = defaultCache,
  define = defaultDefine,
  importmaps = defaultImportMaps,
  compilerOptions = {},
  extraPlugins = [],
  cdnPrefix,
  rollupOptions,
}: CompileOptions) {
  const fallback = createImportMapsFallback({
    importmaps,
    cdnPrefix: cdnPrefix,
  });
  return rollup({
    input,
    plugins: [
      replace({ ...define }),
      workerPlugin(),
      httpResolve({
        cache,
        fallback,
      }),
      virtualFs({ files }),
      transform({
        cdnPrefix,
        compilerOptions: {
          ...defaultCompilerOptions,
          ...compilerOptions,
        },
      }),
      json(),
      ...extraPlugins,
    ],
    ...rollupOptions,
  });
}
