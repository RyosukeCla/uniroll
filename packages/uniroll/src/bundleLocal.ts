import { virtualFs } from "rollup-plugin-virtual-fs";
import replace from "@rollup/plugin-replace";
import workerPlugin from "./plugins/worker-plugin";
import {
  defaultCompilerOptions,
  defaultDefine,
  extractImportSpecfiers,
  transform,
} from "./shared";
import { rollup } from "rollup";
import json from "@rollup/plugin-json";
import { CompileOptions } from "./types";

export function bundleLocal({
  input,
  files,
  define = defaultDefine,
  compilerOptions = defaultCompilerOptions,
  extraPlugins = [],
  rollupOptions,
  cdnPrefix = "https://esm.sh/",
}: CompileOptions) {
  const external = extractImportSpecfiers(Object.values(files))
    .filter((ex) => !ex.startsWith("."))
    .map((p) => `https://esm.sh/${p}`);
  return rollup({
    input,
    external,
    plugins: [
      replace({ ...defaultDefine, ...define }),
      workerPlugin(),
      virtualFs({ files }),
      transform({
        cdnPrefix,
        compilerOptions,
      }),
      json(),
      ...extraPlugins,
    ],
    ...rollupOptions,
  });
}
