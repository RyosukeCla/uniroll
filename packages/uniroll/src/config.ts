export { baseConfigBuilder as base } from "./config/baseConfig";
export { baseConfigBuilderWithTranspile as baseWithTranspile } from "./config/baseConfigWithTranspile";
export { devConfigBuilder as dev } from "./config/devConfig";
export { prodConfigBuilder as prod } from "./config/prodConfig";
