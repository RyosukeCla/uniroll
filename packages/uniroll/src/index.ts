import { Options } from "../index.d";
import { createTransformer } from "./baseTranform";
import { httpResolve } from "rollup-plugin-http-resolve";
import { css } from "rollup-plugin-uniroll-css";
import { transformImportPathToSkypackCDN } from "babel-plugin-transform-import-to-skypack-cdn";
// @ts-ignore
import transformPathToImportMap from "babel-plugin-transform-path-to-import-map";

// @ts-ignore
import ts from "@babel/preset-typescript";
// @ts-ignore
import react from "@babel/preset-react";
// @ts-ignore
import classProperties from "@babel/plugin-proposal-class-properties";
// @ts-ignore
import objectRestSpread from "@babel/plugin-proposal-object-rest-spread";
// @ts-ignore
import nullishCoalescing from "@babel/plugin-proposal-nullish-coalescing-operator";
import path from "path";
import { baseline } from "./baseline";
import {
  createMemoryFs,
  readPkgVersionsIfExists,
  readImportMapIfExists,
} from "./helpers";
import replace from "@rollup/plugin-replace";
import json from "@rollup/plugin-json";

const defaultCache = new Map();

export async function compile(options: Options) {
  const mfs = options.useInMemory ? createMemoryFs(options.files) : options.fs;

  // TODO: Fix me
  const rootpath = options.useInMemory ? "/" : "/";

  const pkgPath = path.join(rootpath, "package.json");
  const importMapPath = path.join(rootpath, "import-map.json");

  const versions = await readPkgVersionsIfExists(mfs, pkgPath);
  const importMap = await readImportMapIfExists(mfs, importMapPath);

  const babelOptions = {
    plugins: [
      classProperties,
      objectRestSpread,
      nullishCoalescing,
      transformPathToImportMap(importMap ?? { imports: {} }),
      transformImportPathToSkypackCDN(
        versions ?? options.versions ?? {},
        (warning) => {
          options.onWarn?.(warning);
        }
      ),
    ],
    presets: [react, ts],
  };

  const baseTransformPlugin = {
    name: "base-transform",
    transform: createTransformer(babelOptions),
  };

  const bundle = baseline({
    ...options,
    fs: mfs,
    rollupPlugins: [
      ...(options.rollupPlugins ?? []),
      replace({ "process.env.NODE_ENV": "development", ...options.replaceMap }),
      json(),
      css(),
      httpResolve({
        cache: options.cache ?? defaultCache,
        onRequest: options.onRequest,
        onUseCache: options.onUseCache,
      }),
      baseTransformPlugin,
    ],
  });
  return bundle;
}
