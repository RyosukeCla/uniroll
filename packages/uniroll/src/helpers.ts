import { ImportMaps } from "rollup-plugin-http-resolve";
import { UnirollConfigBuilderResult } from "./config/base";
import { UnirollOptions } from "./config/base";
import { rollup, RollupOptions } from "rollup";
import type fs from "fs";
import { Volume } from "memfs";
import createFs, { IPromisesAPI } from "memfs/lib/promises";
// import { ImportMap } from "./types";

export function createMemoryFs(files: { [k: string]: string }): IPromisesAPI {
  const vol = Volume.fromJSON(files, "/");
  return createFs(vol) as IPromisesAPI;
}

export type CompileOptions = RollupOptions &
  Omit<UnirollOptions, "fs"> & {
    files: { [k: string]: string };
  };
export function createCompilerOptionBuilder(
  configBuilder: (opts: UnirollOptions) => UnirollConfigBuilderResult
) {
  return async function (opts: CompileOptions) {
    const memfs = createMemoryFs(opts.files);
    const importmaps: ImportMaps | undefined = opts.importMapsPath
      ? JSON.parse((await memfs.readFile(opts.importMapsPath)).toString())
      : undefined;
    const { plugins } = configBuilder({
      fs: memfs,
      importmaps,
      ...opts,
    });
    return {
      ...opts,
      plugins: [...(opts.plugins ?? []), ...plugins],
    };
  };
}

export function createCompiler(
  getConfig: (opts: UnirollOptions) => UnirollConfigBuilderResult
) {
  const buildConfig = createCompilerOptionBuilder(getConfig);
  return async function (opts: CompileOptions) {
    const config = await buildConfig(opts);
    return rollup(config);
  };
}

export async function readPkgVersionsIfExists(
  fs_: typeof fs.promises,
  pkgPath: string
): Promise<{ [key: string]: string } | null> {
  try {
    const pkgStr = await fs_.readFile(pkgPath, "utf-8");
    const pkg = JSON.parse(pkgStr);
    return { ...pkg?.dependencies };
  } catch (err) {
    return null;
  }
}

export async function readImportMapIfExists(
  fs_: typeof fs.promises,
  importMapPath: string
): Promise<{ imports: { [key: string]: string } } | null> {
  try {
    const importMapStr = await fs_.readFile(importMapPath, "utf-8");
    return JSON.parse(importMapStr);
  } catch (err) {
    return null;
  }
}
