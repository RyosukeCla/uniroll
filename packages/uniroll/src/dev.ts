import { createCompiler, createMemoryFs } from "./helpers";
import { dev } from "./config/devConfig";
export const compile = createCompiler(dev);
export { createMemoryFs };
