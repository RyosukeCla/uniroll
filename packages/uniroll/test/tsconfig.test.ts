import "isomorphic-unfetch";

import { compile } from "../src/index";
const files = {
  "/import-map.json": JSON.stringify({
    imports: {},
  }),
  "/index.tsx": "const x = 1;console.log(x)",
};

test("transform with tsconfig", async () => {
  try {
    const bundle = await compile({
      files,
      input: "/index.tsx",
      tsconfig: {
        target: "es5",
      },
      onwarn() {},
    });
    const out = await bundle.generate({ format: "es" });
    // console.log(out.output[0].code);
    expect(out.output[0]).toMatchSnapshot();
  } catch (err) {
    console.log(err);
    throw err;
  }
});
