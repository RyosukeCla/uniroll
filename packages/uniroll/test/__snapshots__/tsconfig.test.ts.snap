// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform with tsconfig 1`] = `
Object {
  "code": "const x = 1;
console.log(x);
",
  "dynamicImports": Array [],
  "exports": Array [],
  "facadeModuleId": "/index.tsx",
  "fileName": "index.js",
  "implicitlyLoadedBefore": Array [],
  "importedBindings": Object {},
  "imports": Array [],
  "isDynamicEntry": false,
  "isEntry": true,
  "isImplicitEntry": false,
  "map": null,
  "modules": Object {
    "/index.tsx": Object {
      "originalLength": 26,
      "removedExports": Array [],
      "renderedExports": Array [],
      "renderedLength": 28,
    },
  },
  "name": "index",
  "referencedFiles": Array [],
  "type": "chunk",
}
`;
